Pooled memory allocator

MoloToha opened the pull request:
Uploaded basic prototype for pooled memory allocator along with a very basic test to just see if it works at all. I will try to write a more complex test and resolve TODO's in the nearest future.

rafalcieslak commented on prototypes/palloc/pool_alloc.c:
```diff
@@ -0,0 +1,212 @@
+int page_size;
+
+typedef struct pool_item {
+    uint64_t pi_guard_number; //0xcafebabecafebabe by default, normally isn't changed, so if it has another value, memory corruption has taken place
```
When picking magic guard values, it is a good practice to pick one that does not contain any (long) substring more than once. The one you picked is particularly unfortunate: `cafebabe` repeats twice at a word-boundary! So if one happens to write this value to address `X` and afterwards at `X+4`, they overlay perfectly, and it may seem that `X` still holds a valid guard number, even though it was overwritten.

rafalcieslak commented on prototypes/palloc/pool_alloc.c:
```diff
@@ -0,0 +1,212 @@
+    slab -> ph_start = sizeof(pool_slab_t)+align_to_binary_pow(slab -> ph_ntotal, 3);
+    memset(slab -> ph_bitmap, 0, slab -> ph_ntotal);
+    for (int i = 0; i < slab -> ph_ntotal; i++) {
+        pool_item_t *curr_pi = (pool_item_t*) ((uint64_t) slab+(slab -> ph_start)+i*(size+sizeof(pool_item_t)));
```
You perform this computation (slab address to pool_item address) many times - consider wrapping it in a convenience macro, it would increase readability and decrease the chance of mistakes.

rafalcieslak commented on prototypes/palloc/pool_alloc.c:
```diff
@@ -0,0 +1,212 @@
+        } else {
+            new_slab = LIST_FIRST(&(pool -> pp_part_slabs));
+            LIST_REMOVE(new_slab, ph_slablist);
+        }
```
These two `if` cases have almost identical content. Maybe it would be simpler to write:

```c
pool_slab_list_t* slab_list_to_use = pool->pp_nitems ? &pool->pp_empty_slabs : &pool->pp_part_slabs;
new_slab = LIST_FIRST(slab_list_to_use);
LIST_REMOVE(new_slab, ph_slablist);
```
or maybe even:
```c
new_slab = LIST_FIRST(pool->pp_nitems ? &pool->pp_empty_slabs : &pool->pp_part_slabs);
LIST_REMOVE(new_slab, ph_slablist);
```
Similar transformation can be used to simplify many of your `if` statements - but I am not sure if the above suggestion fits our coding style.

MoloToha replied:
I usually write this kind of code for readability purposes. We can discuss this at the meeting.

rafalcieslak commented on prototypes/palloc/pool_alloc.c:
```diff
@@ -0,0 +1,212 @@
+    }
+
+    for (size_t i = 0; i < sizeof(pool_t)/8; i++) {
+        tmp[i] = 0xdeadbeefdeadbeef;
```
Am I correct that this causes writes to memory regions that were already freed? It might be better to first perform these writes (e.g. while checking the `tmp[i]` values at the beginning of `pool_destroy`), and destroy slabs afterwards.

MoloToha replied:
This piece of code writes to a pool_t object which is located at client's side (so it can't be freed for the second time), slabs are destroyed separately before this operation.

rafalcieslak commented:
Also: remember to run `make format` periodically - it will help you clean up your code.

Additionally, you can skip some parenthesis:

`(x->y)++` is the same as `x->y++`,

`&(x->y)` is the same as `&x->y`.

MoloToha commented:
Thank you for suggestions! I will implement them ASAP.

cahirwpz commented:
@MoloToha Please address code changes outlined by @rafalcieslak. The suggestions can be viewed by clicking *show outdated* on this page. If I hadn't run `clang-format` they wouldn't be hidden – sorry for inconvenience.

MoloToha commented:
The last commit is my attempt to integrate this component into kernel, everything works (at least on my computer), test runs correctly, currently this test is set up to cause memory corruption, so you should comment line containing memset (which causes this memory corruption) in order to pass the test.

rafalcieslak commented:
I see this is shaping up nicely. Note that throughout the codebase we prefer:
```c
/* C-style comments */
```
over:
```c++
// C++-style comments.
```

If you want to prepare a test which intentionally crashes (to demonstrate something), you can pass `KTEST_FLAG_BROKEN` to `KTEST_ADD` - this way the test will not be launched automatically during automatic testing, but you will still be able to run it by specifying `test=...`.

MoloToha commented:
This memset writes zeroes to space that is occupied by other items inside a slab, thus creating an artificial memory corruption

rafalcieslak commented on tests/pool_alloc_corruption.c:
```diff
@@ -0,0 +1,49 @@
+void another_int_destr(__unused void *buf, __unused size_t size) {
+}
+
+int test_pool_alloc_corruption() {
```
You can avoid code duplication by merging this function into one: `test_pool_alloc(int corrupt_memory)` and then using both variants from proxy functions:
```
int test_pool_alloc_corruption(){
  return test_pool_alloc(1);
} ... etc.
```

MoloToha replied:
Done

rafalcieslak commented on sys/pool_alloc.c:
```diff
@@ -0,0 +1,235 @@
+#define PALLOC_DEBUG 0
+
+#if defined(PALLOC_DEBUG) && PALLOC_DEBUG > 0
+#define PALLOC_DEBUG_PRINT(text, args...)                                      \
```
We don't need yet another macro for displaying messages - just use `log`. I see you wanted to quickly enable/disable pool allocator debug messages and that is indeed useful, but #239 (klog) should be ready very soon and it will enable kernel-wide message filtering. Also, generally, your debug messages are very verbose. Keeping track of the general state of the allocator is great, but displaying messages on every single allocation seems like a bit too much. Most of your messages seem to track control flow ("Entering..."), it it simpler to track it using breakpoints in a debugger. Now that you've tested the allocator and it seems to be working okay, you may want to trim down these messages a bit.

MoloToha replied:
Okay, I will change this as soon as klog is ready

rafalcieslak commented on sys/pool_alloc.c:
```diff
@@ -0,0 +1,235 @@
+  (pool_item_t *)((unsigned long)slab + (slab->ph_start) +                     \
+                  (i) * (size + sizeof(pool_item_t)))
+
+#define ALIGN_TO_BINARY_POW(x, a) ((((x) >> (a)) + 1) << (a))
```
Correct me if I'm wrong, but I think that if `x` is already aligned, then this macro will return a needlessly larger value. Also you may want to use `align` from `common.h`.

rafalcieslak commented on sys/pool_alloc.c:
```diff
@@ -0,0 +1,235 @@
+} pool_item_t;
+
+pool_slab_t *create_slab(size_t size, pool_t *pool,
+                         void (*constructor)(void *, size_t)) {
```
Is the `constructor` argument necessary? The constructor is stored within `pool`, so it seems redundant.

rafalcieslak commented on include/pool_alloc.h:
```diff
@@ -0,0 +1,34 @@
+  uint16_t ph_ntotal; /* total number of chunks*/
+  uint16_t ph_start;  /* start offset in page */
+  bitstr_t ph_bitmap[0];
+} pool_slab_t;
```
The definition of the contents of this structure is only needed inside `pool_alloc.c`. Moving the structure definition there will make it invisible from other files, which will clean up the interface a bit, and will prevent people from accessing slab metadata.

rafalcieslak commented on sys/pool_alloc.c:
```diff
@@ -0,0 +1,235 @@
+  slab->ph_nfree = slab->ph_ntotal;
+  slab->ph_start =
+    sizeof(pool_slab_t) + ALIGN_TO_BINARY_POW(slab->ph_ntotal, 2);
+  memset(slab->ph_bitmap, 0, slab->ph_ntotal);
```
I think you are clearing too much with this memset, the size of a bitmap with n entries is `bitstr_size(n)`.

rafalcieslak commented on sys/pool_alloc.c:
```diff
@@ -0,0 +1,235 @@
+      take it (no construction required);
+  else {
+      allocate memory;
+      construct the object;
```
Is this comment accurate? No object construction happens in this function. Maybe you meant slab creation? Anyway, *interleaving* the code with comments makes it simpler to read than providing the outline at the beginning of some functions. This way, when one reads thorough the code, the comments explain which part of the code is responsible for which part of your logic. A general outline before the actual code is only helpful if the code is very complex and the outline significantly simplifies the code.

rafalcieslak commented on tests/pool_alloc.c:
```diff
@@ -0,0 +1,70 @@
+#include <malloc.h>
+#include <ktest.h>
+
+void int_constr(void *buf, __attribute__((unused)) size_t size) {
```
`common.h` provides `__unused`.

rafalcieslak commented on tests/pool_alloc.c:
```diff
@@ -0,0 +1,70 @@
+  *num = 0;
+}
+void int_destr(__attribute__((unused)) void *buf,
+               __attribute__((unused)) size_t size) {
```
I have a feeling it will be common for many simple objects allocated with the pool allocator to have no destructor. Well, maybe. In that case I would need to provide such pointless destructor function just to satisfy the interface. It might be convenient to have a no-op `pool_default_destructor` defined in `pool_alloc.c` so that one might:
```c
pool_init(&my_pool, size, my_constr, pool_default_destructor);
```
without creating additional empty destructor. We do a similar thing with `vnode` operations: a number of trivial `vnode_generic_*` is provided so that one does not have to write their own empty functions.

rafalcieslak commented on sys/pool_alloc.c:
```diff
@@ -0,0 +1,235 @@
+  unsigned long pi_data[0];
+} pool_item_t;
+
+pool_slab_t *create_slab(size_t size, pool_t *pool,
```
`create_slab`, `destroy_slab` and `slab_alloc` are not a part of the interface, right? Make sure to mark these functions as `static`, so that they are inaccessible from other files.

rafalcieslak commented on sys/pool_alloc.c:
```diff
@@ -0,0 +1,235 @@
+  return p;
+}
+void pool_free(pool_t *pool,
+               void *ptr) { /* TODO: implement empty slab management */
```
Remember that `TODO` comments must stay informative for a long time. Some (decent!) empty slab management is already implemented, so this comment does not tell me what exactly is missing. Imagine it's year 2020, you don't remember any of this code, and someone tasked you with resolving this `TODO` :) Could you make it a bit more verbose?

rafalcieslak commented on tests/pool_alloc.c:
```diff
@@ -0,0 +1,70 @@
+}
+
+void test_pool_alloc(
+  int flag) { /* 0 - regular test, 1 - memory corrruption, 2 - double free */
```
I know, I know, this is just a simple test - but these flags would be much simpler to read and understand if you made them a `#define` or `enum`.

rafalcieslak commented:
So I gave this branch a close examination and (besides my previous comments) I believe this implementation is sound and solid. Good job :+1:! As you see, most of my remarks are about code style - I really don't have much to complain about how the pool allocator functions. I believe this branch will be ready for merging very soon, although certainly @cahirwpz will have much more stylistic remarks than I do.

MoloToha commented:
I addressed all issues, the last one remaining is proper calculation of ph_ntotal (total # of items in slab), because atm it is still done by the old formula which uses size of the old bitmap (8 times bigger than the current one) and uses the worst case value for bitmap padding (which is 3 bytes).

cahirwpz commented on sys/pool.c:
```diff
@@ -0,0 +1,203 @@
+}
+
+static pool_slab_t *create_slab(pool_t *pool) {
+  /* log("create_slab: pool = %p, pp_itemsize = %d", pool, pool->pp_itemsize);
```
Instead of commenting it out I suggest to use `debug` macro:
```c
#ifdef DEBUG
#define debug(...) log(__VA_ARGS__)
#else
#define debug(...)
#endif

cahirwpz commented on sys/pool.c:
```diff
@@ -0,0 +1,203 @@
+  for (int i = 0; i < slab->ph_ntotal; i++) {
+    pool_item_t *curr_pi = get_pi_at_idx(slab, i, pool->pp_itemsize);
+    if (pool->pp_dtor)
+      pool->pp_dtor(curr_pi->pi_data, pool->pp_itemsize);
```
Is it necessary to pass `pool->pp_itemsize` here and in other places? What if it the value was present in `pool_slab` structure as well?

cahirwpz commented on sys/pool.c:
```diff
@@ -0,0 +1,203 @@
+  slab->ph_nused = 0;
+  /* sizeof(pool_slab_t) + n * (sizeof(pool_item_t) + size)
+   * + ((n + 31) / 32) * 4 <= PAGESIZE */
+  slab->ph_ntotal = (8 * (PAGESIZE - sizeof(pool_slab_t)) - 31) /
```
Could you think of expressing it in a simpler way?

cahirwpz commented:
We're almost there... I'll have another detailed read this week and apply some cosmetic changes.

rafalcieslak commented:
I am not 100% sure, but it appears as though the last [failed test result](https://travis-ci.org/cahirwpz/mimiker/builds/225681324) was due to spurious qemu lag. Unless we keep experiencing this particular problem (deadlock after ` Sleep 'main' thread on 'callout_good'`), I suggest we ignore it - the callouts interface will be getting a major upgrade soon anyway.

cahirwpz commented:
@MoloToha Please do not modify this PR anymore if not requested. I'm taking it over in order to clean it up for merge with master.
